DependencyInjection/Compiler/EntityListenerPass.php

```diff
@@ -39,20 +39,50 @@ public function process(ContainerBuilder $container)
                     continue;
                 }
 
-                $resolver = sprintf('doctrine.orm.%s_entity_listener_resolver', $name);
-                if ($container->hasAlias($resolver)) {
-                    $resolver = (string) $container->getAlias($resolver);
+                $resolverId = sprintf('doctrine.orm.%s_entity_listener_resolver', $name);
+                if ($container->hasAlias($resolverId)) {
+                    $resolverId = (string) $container->getAlias($resolverId);
                 }
 
-                if (!$container->hasDefinition($resolver)) {
+                if (!$container->hasDefinition($resolverId)) {
                     continue;
                 }
 
+                $resolver = $container->getDefinition($resolverId);
+
                 if (isset($attributes['entity']) && isset($attributes['event'])) {
                     $this->attachToListener($container, $name, $id, $attributes);
                 }
 
-                $container->getDefinition($resolver)->addMethodCall('register', array(new Reference($id)));
+                if (isset($attributes['lazy']) && $attributes['lazy']) {
+                    $listener = $container->findDefinition($id);
+
+                    if (!$listener->isPublic()) {
+                        throw new \InvalidArgumentException(sprintf('The service "%s" must be public as this entity listener is lazy-loaded.', $id));
+                    }
+
+                    if ($listener->isAbstract()) {
+                        throw new \InvalidArgumentException(sprintf('The service "%s" must not be abstract as this entity listener is lazy-loaded.', $id));
+                    }
+
+                    $interface = 'Doctrine\\Bundle\\DoctrineBundle\\Mapping\\EntityListenerServiceResolver';
+                    $class = $resolver->getClass();
+
+                    if (substr($class, 0, 1) === '%') {
+                        // resolve container parameter first
+                        $class = $container->getParameter(trim($class, '%'));
+                    }
+
+                    if (!is_a($class, $interface, true)) {
+                        throw new \InvalidArgumentException(
+                            sprintf('Lazy-loaded entity listeners can only be resolved by a resolver implementing %s.', $interface)
```

Comment:
I think not: if you mark a listener as lazy you specifically opt-in to certain behaviour. Silently falling back on a different solution will not only change the expected behaviour but most likely trigger a fatal error since most of the time you'll want to use this with a service that has a constructor with arguments.